python/mxnet/contrib/nn/layer.py

```diff
@@ -0,0 +1,355 @@
+# coding: utf-8
+# pylint: disable= arguments-differ
+"""Neural network layers."""
+
+from ... import symbol, ndarray
+from ...symbol import Symbol
+from ...ndarray import NDArray
+from .parameter import Parameter, ParameterDict
+
+
+class Layer(object):
+    """Base class for all neural network layers and models.
+
+    Your models should subclass this class.
+
+    Layers can also contain other Layers, allowing you to nest them in a tree
+    structure. You can assign sublayers as regular attributes::
+        from mxnet import nn
+        class Net(nn.Layer):
+            def __init__(self, prefix=None, params=None):
+                super(Net, self).__init__(prefix=prefix, params=params)
+                self.dense1 = nn.Dense(20, in_units=10, prefix=self.prefix+'dense1_')
+                self.dense2 = nn.Dense(20, in_units=20, prefix=self.prefix+'dense2_')
+
+            def forward(self, x):
+                x = self.dense1(x)
+                return self.dense2(x)
+
+    Sublayers assigned this way will be registered and will have their status changed
+    too when you call .train() etc.
+
+    Parameters
+    ----------
+    prefix : str
+        Prefix acts like a name space. It will be prepended to the name of all Symbols and
+        Parameters created by this layer. Prefix should be unique within one network
+        to prevent name collisions.
+    params : ParameterDict or None
+        Manages Parameters of this Layer and sublayers. You can make two Layers share
+        parameter by passing the same dictionary to them. For example::
+            params = nn.ParameterDict(prefix='dense_')
+            dense1 = nn.Dense(20, in_units=10, prefix='dense1_', params=params)
+            dense2 = nn.Dense(20, in_units=10, prefix='dense2_', params=params)
+
+        dense1 and dense2 now have shared weights.
+    """
+    _counter = {}
+
+    def __init__(self, prefix=None, params=None):
+        self._prefix = self._make_prefix(prefix)
+        self._name = self.prefix[:-1] if self.prefix.endswith('_') else self.prefix
+        if params is None:
+            params = ParameterDict(prefix=self.prefix)
+        self._params = params
+        self._children = []
+
+    def __setattr__(self, name, value):
+        """Automatically register sublayers."""
+        super(Layer, self).__setattr__(name, value)
```

Comment:
Chainer used to use explicit registration at `__init__` time, and is moving to PyTorch's implicit registration on assign because it works much better with IDEs.